com.sap.cloud.lm.sl.cf.process/src/test/java/com/sap/cloud/lm/sl/cf/process/steps/DetermineDesiredStateAchievingActionsStepTest.java

```diff
@@ -1,177 +1,420 @@
-        context.setVariable(Constants.VAR_APP_CONTENT_CHANGED, Boolean.toString(hasAppChanged));
-        context.setVariable(Constants.VAR_VCAP_APP_PROPERTIES_CHANGED, hasAppPropertiesChanged);
-        context.setVariable(Constants.VAR_VCAP_SERVICES_PROPERTIES_CHANGED, hasServicesPropertiesChanged);
-        context.setVariable(Constants.VAR_USER_PROPERTIES_CHANGED, hasUserPropertiesChanged);
-        context.setVariable(Constants.PARAM_NO_START, false);
-        when(appStateCalculator.computeCurrentState(any())).thenReturn(currentAppState);
-        when(appStateCalculator.computeDesiredState(any(), eq(false))).thenReturn(desiredAppState);
-        step.appStateCalculatorSupplier = () -> appStateCalculator;
-        CloudApplicationExtended app = new CloudApplicationExtended(null, "dummy");
-        app.setRestartParameters(appRestartParameters);
-        context.setVariable(Constants.VAR_APP_TO_DEPLOY, JsonUtil.toJson(app));
+    public DetermineDesiredStateAchievingActionsStepTest(ApplicationStartupState currentAppState, ApplicationStartupState desiredAppState,
+        boolean hasAppChanged, Set<ApplicationStateAction> expectedAppStateActions, List<CloudBuild> cloudBuilds,
+        Class<? extends Exception> exception) {
+        super(currentAppState, desiredAppState, hasAppChanged, expectedAppStateActions, cloudBuilds);
+        this.exception = exception;
     }
 
     @Test
     public void testExecute() throws Exception {
+        if (exception != null) {
+            expectedException.expect(exception);
+        }
         step.execute(context);
-
         assertStepFinishedSuccessfully();
 
         assertEquals(expectedAppStateActions, StepsUtil.getAppStateActionsToExecute(context));
     }
 
     @Override
-    protected DetermineDesiredStateAchievingActionsStep createStep() {
-        return new DetermineDesiredStateAchievingActionsStep();
+    protected void initProperties() {
+        context.setVariable(Constants.VAR_VCAP_APP_PROPERTIES_CHANGED, false);
+        context.setVariable(Constants.VAR_VCAP_SERVICES_PROPERTIES_CHANGED, false);
+        context.setVariable(Constants.VAR_USER_PROPERTIES_CHANGED, false);        
     }
 
+    @Override
+    protected RestartParameters getRestartParameters() {
+        return new RestartParameters(false, false, false);
+    }
+
+    @RunWith(Parameterized.class)
+    public static class DetermineAppRestartTest extends DetermineDesiredStateAchievingActionsStepBaseTest {
+
+
+        @Parameters
+        public static List<Object[]> getParameters() throws ParseException {
+            return Arrays.asList(new Object[][] {
+                // @formatter:off
+                // (0)
+                {
+                    ApplicationStartupState.STARTED, ApplicationStartupState.STARTED, false, Stream.of(ApplicationStateAction.STAGE, ApplicationStateAction.START, ApplicationStateAction.STOP).collect(Collectors.toSet()), Collections.emptyList(), true, true, true, true, true, true
```

Comment:
The current and desired states in this test are always `ApplicationStartupState.STARTED`. This means that they are not actual parameters but constants. Extract them as such.

Also, the `hasAppChanged` field is also always `false`. Extract it as well.